### DISCUSSION TITLE
Is there a way to build multiple doc-sites in single repo which are not merged into a single doc set?

### DISCUSSION THREAD

**Discussion by RiyaTyagi** [mkdocs/mkdocs#2412] [Q&A]
I want to use a single repository, which is shared by multiple teams, to store multiple doc sets but I don't want them to be merged into a single doc set with shared navigation which is the case with monorepo plugin. The individual doc set should not be aware that the other doc sets exist. For example:
```terminal
├── mkdocs.yml
├── set1
│   ├── docs
│   │   ├── index.md
│   │   └── test1.md
│   └── mkdocs.yml
└── set2
    ├── docs
    │   ├── index.md
    │   └── test1.md
    └── mkdocs.yml
```
By providing some config value at root mkdocs.yml after the build I should get something like below in the site folder, or a variation of it to include root index.html which contains links to individual site,  without having to run the build command for each sub-directory:

```terminal
├── set1
│   ├── 404.html
│   ├── assets
│   ├── index.html
│   ├── search
│   │   └── search_index.json
│   ├── sitemap.xml
│   ├── sitemap.xml.gz
│   └── test1
│       └── index.html
└── set2
    ├── 404.html
    ├── assets
    ├── index.html
    ├── search
    │   └── search_index.json
    ├── sitemap.xml
    ├── sitemap.xml.gz
    └── test1
        └── index.html

```
Here, doc set **set1** is not aware that **set2** exists.

I am not sure if such a plugin or config already exists but if not I would like to create a new plugin for it.

**Comment by martinbira** [comment]
Perhaps https://github.com/backstage/mkdocs-monorepo-plugin or https://github.com/wilhelmer/mkdocs-multirepo is what you're looking for.

**Reply by RiyaTyagi** [reply]
I have looked into both of them the mkdocs-monorepo-plugin merges the doc sets together, which is not what I am looking for, and mkdocs-multirepo does not merge the doc sets but fetches the data from different repositories/ git submodules. I am looking for a kind of a combined solution i.e, a monorepo plugin without merging the doc sets into one.

**Comment by waylan** [comment]
Each "doc set" would need its own config file and be its own standalone "site". When running MkDocs' commands you can specify the config file. As all paths in a config file are relative to the config file (not the current working directory) you could arrange things however you like (perhaps have each site write out to a subdir of the same parant dir). If you want to build all sites at once, a simple script should be able to step through all of them and build each in turn.

**Reply by felipoantonoff** [reply]
Good idea, maybe using https://www.mkdocs.org/user-guide/configuration/#configuration-inheritance and a small bash or python script to build all projects.

**Comment by felipoantonoff** [comment]
Interesting idea, I've been thinking about something like that too.

A general mkdocs.yml and within each subfolder/project has its own with some changes.

And with a build command you can generate the HTML for each project, great for those who want to have N sites, but keep the central repository to facilitate adjustments.

**Comment by kirannhegde** [comment]
Hello @RiyaTyagi @waylan  @FelipoAntonoff @martinbira 

I am currently trying to solve an issue described at: https://github.com/backstage/mkdocs-monorepo-plugin/issues/71
. Here is the issue described again:

> We make use of a plugin called mkdocstrings to generate documentation for the source code. This plugin loads the different Python libraries mentioned in the source code during the documentation generation process. As a result of this, there are Python dependency resolution issues that we encounter when using the monrepo plugin in a root repository . If the mkdocs.yaml file in the root repository references multiple mkdocs.yaml files from different repositories, we run into dependency resolution errors as there are conflicting dependency requirements in the different repositories.
> 
> The code in each of the different repositories uses different versions of the same Python module. Hence, the issue.
> I am wondering that this must be an issue at a lot of organizations.
> How is this overcome elsewhere?
> 
> To overcome this, we were thinking that the docs folder in each of the different repositories should be processed independently and the output of the mkdocs build command should be copied to some place in the root repository and then have mkdocs somehow assemble the documentation. I am not sure if this is even possible. If someone has already done this, please respond back.

What i would like to know is:
-If the search indexes created(search_index.json) from the different builds of mkdocs in the various repositories can be combined easily?
If so, would this be a safe approach to have search working in the main site?
-Also, is there a mkdocs plugin which already does the task of combining the various search indexes?

Regards,
Kiran Hegde

**Comment by ssbarnea** [comment]
After trying both mentioned plugins and trying to find workaround for their bugs/limitations I would say that currently mkdocs is (mostly) not suitable for combining multiple sites.

The main issue with it is that it it seems impossible to have any support for includes (like snippets), so paths willbe broken as mkdocs paths are relative to main config file (standalone site vs consolidated one will have mkdocs in different locations on disk).

Basically it seems impossible to make it work unless mkdocs really gets support for true-relative paths, relative to current file.

**Reply by matthiasPOE** [reply]
Did you find a working alternative?

**Reply by squidfunk** [reply]
It should be achievable with the [projects plugin](https://squidfunk.github.io/mkdocs-material/plugins/projects/) (from [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/)) and [`project_config_transform`](https://squidfunk.github.io/mkdocs-material/plugins/projects/#config.projects_config_transform). For instance, see [the transform we're using](https://github.com/mkdocs-material/examples/blob/master/examples/__init__.py) to build our [examples](https://github.com/mkdocs-material/examples) repository, which sets some of the paths for nested projects. As many other of our built-in plugins, the projects plugin was born because of MkDocs does not provide the functionality.

**Comment by sarathTegra** [comment]
Incase anyone stumbles upon this, the [mkdocs-multirepo](https://github.com/wilhelmer/mkdocs-multirepo) can handle this use-case, if you just input the folder name of the sets in the multirepo config.yml file and only run the **mkdocs-multirepo --build** command.

**Comment by KonstantinNosov** [comment]
I had similar task but rather wanted to have docs to be in each repo and build it as packages with version synced with repo artifact.
So I've developed plugin for mkdocs and toolset to pack individual projects docs and site (mkdocs.yaml + customization) to python packages and at static html  build time I install site package and those project packages that are needed to venv and build with mkdocs. 

Base plugin used to build project docs is also capable to inject any dir out of `docs_dir ` to site structure.

Here is my solution [documentation](https://docs.exordis.com/documentation/partial-documentation/) with use cases and links to tooling

**Comment by DrewScatterday** [comment]
If anyone finds this from google I did something like this with github actions 

```yaml
name: Build & Deploy MkDocs

on:
  push:
    branches:
      - main
      - develop

permissions:
  contents: write 
  pages: write       
  id-token: write    

jobs:
  build-and-deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          persist-credentials: false

      - name: Setup Python
        uses: actions/setup-python@v5
        with:
          python-version: '3.13'

      - name: Install dependencies
        run: pip install -r requirements.txt 

      - name: Build MkDocs
        run: mkdocs build --clean

      - name: Upload Pages Artifact
        if: github.ref == 'refs/heads/main'
        uses: actions/upload-pages-artifact@v3
        with:
          path: ./site

      - name: Deploy Production to GitHub Pages
        if: github.ref == 'refs/heads/main'
        id: deploy
        uses: actions/deploy-pages@v4

      - name: Deploy Dev to External Repo
        if: github.ref == 'refs/heads/develop'
        uses: peaceiris/actions-gh-pages@v4
        with:
          personal_token: ${{ secrets.DEV_SITE_DEPLOY_TOKEN }}
          external_repository: user/my-develop-docsite 
          publish_branch: main
          publish_dir: ./site
          destination_dir: docs
          user_name: 'github-actions[bot]'
          user_email: 'github-actions[bot]@users.noreply.github.com'
```

If someone pushes to main - it pushes to the base repo github pages site.

If someone pushes to develop - it pushes the code to a separate develop repo which I have preconfigured with github pages to point at the /docs directory. It sucks to have two repos but this works. Another option is using netlify for the dev site and github pages for main

it requires a personal access token for github actions to push to the other repo
